src/puppetlabs/puppetdb/utils/string_formatter.clj

```diff
@@ -0,0 +1,101 @@
+
+(defn re-quote
+  "Quotes s so that all of its characters will be matched literally."
+  [s]
+  ;; Wrap all segments not containing a \E in \Q...\E, and replace \E
+  ;; with \\E.
+  (apply str
+         "\\Q"
+         (concat (->> (string/split s #"\\E" -1)
+                      (interpose "\\E\\\\E\\Q"))
+                 ["\\E"])))
+
+(defn maybe-strip-escaped-quotes
+  [s]
+  (if (and (> (count s) 1)
+           (string/starts-with? s "\"")
+           (string/ends-with? s "\""))
+    (subs s 1 (dec (count s)))
+    s))
+
+(defn quoted
+  [s]
+  (str "'" s "'"))
+
+(defn comma-separated-keywords
+  [words]
+  (let [quoted-words (map quoted words)]
+    (if (> (count quoted-words) 2)
+      (str (string/join ", " (butlast quoted-words)) ", " "and " (last quoted-words))
+      (string/join " and " quoted-words))))
+
+(defn dashes->underscores
+  "Accepts a string or a keyword as an argument, replaces all occurrences of the
+  dash/hyphen character with an underscore, and returns the same type (string
+  or keyword) that was passed in.  This is useful for translating data structures
+  from their wire format to the format that is needed for JDBC."
+  [str]
+  (let [result (string/replace (name str) \- \_)]
+    (if (keyword? str)
+      (keyword result)
+      result)))
+
+(defn underscores->dashes
+  "Accepts a string or a keyword as an argument, replaces all occurrences of the
+   underscore character with a dash, and returns the same type (string
+   or keyword) that was passed in.  This is useful for translating data structures
+   from their JDBC-compatible representation to their wire format representation."
+  [s]
+  (let [opt-key? (optional-key? s)
+        result (if opt-key?
+                 (string/replace (name (:k s)) \_ \-)
+                 (string/replace (name s) \_ \-))]
+    (cond
+      opt-key? (if (keyword? (:k s))
+                 (s/optional-key (keyword result))
+                 (s/optional-key result))
+      (keyword? s) (keyword result)
+      :else result)))
+
+(defn dash->underscore-keys
+  "Converts all top-level keys (including nested maps) in `m` to use dashes
+  instead of underscores as word separators"
+  [m]
+  (sp/transform [sp/ALL]
+                #(update % 0 dashes->underscores)
+                m))
+
+(defn underscore->dash-keys
+  "Converts all top-level keys (including nested maps) in `m` to use underscores
+  instead of underscores as word separators"
+  [m]
+  (sp/transform [sp/ALL]
+                #(update % 0 underscores->dashes)
+                m))
+
+(defn pprint-json-parse-exception
+  "Returns a parsed JsonParseException message.
+  From the second line of the error message, the line and
+  column index are extracted, so that we can place an arrow,
+  pointing at that position. "
+  [error query]
+  (let [error-lines (string/split-lines (.getMessage error))
+        [_ line column] (re-find #"line: (\d+), column: (\d+)" (second error-lines))
+        line-index (Integer. line)
```

Comment:
Probably the conversion to Integer and the subtraction can be performed nicer, but I don't know how.

Reply:
An alternative to the integer constructor:  https://github.com/puppetlabs/clj-kitchensink/blob/main/src/puppetlabs/kitchensink/core.clj#L208-L216 but will return nil.  
I believe if the parse fails now, it will throw an exception.  I wonder what is the user visible output in that case.